crates/snark-acct-types/src/outputs.rs

```diff
@@ -1,90 +1,122 @@
+    /// Creates new update outputs.
     pub fn new(transfers: Vec<OutputTransfer>, messages: Vec<OutputMessage>) -> Self {
         Self {
-            transfers,
-            messages,
+            transfers: transfers.into(),
+            messages: messages.into(),
         }
     }
 
+    /// Creates empty update outputs.
     pub fn new_empty() -> Self {
         Self::new(Vec::new(), Vec::new())
     }
 
+    /// Gets the transfers.
     pub fn transfers(&self) -> &[OutputTransfer] {
         &self.transfers
     }
 
-    pub fn transfers_mut(&mut self) -> &mut Vec<OutputTransfer> {
+    /// Gets mutable transfers.
+    pub fn transfers_mut(
+        &mut self,
+    ) -> &mut VariableList<OutputTransfer, { MAX_TRANSFERS as usize }> {
         &mut self.transfers
     }
 
+    /// Gets the messages.
     pub fn messages(&self) -> &[OutputMessage] {
         &self.messages
     }
 
-    pub fn messages_mut(&mut self) -> &mut Vec<OutputMessage> {
+    /// Gets mutable messages.
+    pub fn messages_mut(&mut self) -> &mut VariableList<OutputMessage, { MAX_MESSAGES as usize }> {
         &mut self.messages
     }
-}
 
-/// Transfer from one account to another.
-///
-/// This IS NOT a message and DOES NOT carry data.  This is NOT intended to be
-/// for sending funds between subjects.  This is primarily intendede for future
-/// functionality related to sequencer accounts.
-#[derive(Copy, Clone, Debug, Eq, PartialEq)]
-pub struct OutputTransfer {
-    dest: AccountId,
-    value: BitcoinAmount,
+    /// Tries to extend transfers with items from an iterator.
+    ///
+    /// Returns an error if adding all items would exceed capacity.
+    /// Does not modify the list if capacity would be exceeded.
+    pub fn try_extend_transfers<I>(&mut self, iter: I) -> Result<(), &'static str>
+    where
+        I: IntoIterator<Item = OutputTransfer>,
+    {
+        let items: Vec<_> = iter.into_iter().collect();
+        let needed = self.transfers.len() + items.len();
+
+        if needed > MAX_TRANSFERS as usize {
+            return Err("transfers capacity would be exceeded");
+        }
+
+        for item in items {
+            self.transfers.push(item).expect("capacity already checked");
+        }
+
+        Ok(())
+    }
+
+    /// Tries to extend messages with items from an iterator.
+    ///
+    /// Returns an error if adding all items would exceed capacity.
+    /// Does not modify the list if capacity would be exceeded.
+    pub fn try_extend_messages<I>(&mut self, iter: I) -> Result<(), &'static str>
+    where
+        I: IntoIterator<Item = OutputMessage>,
+    {
+        let items: Vec<_> = iter.into_iter().collect();
+        let needed = self.messages.len() + items.len();
+
+        if needed > MAX_MESSAGES as usize {
+            return Err("messages capacity would be exceeded");
+        }
+
+        for item in items {
+            self.messages.push(item).expect("capacity already checked");
+        }
+
+        Ok(())
+    }
```

Comment:
Well this isn't that useful, if we were going to collect it into a vec anyways we'd just accept a vec, or really a slice.  But actually we'd want to use an `ExactSizeIterator`.  But this is something we can worry about later.